### DISCUSSION TITLE
Changing Default Hash Implementation from CRC32 to CRC-CCITT

### DISCUSSION THREAD

**Discussion by ebenezerarunkumar** [nasa/fprime#1639] [Q&A]
Hi,

For my project, I want to change the standard hashing implementation in fprime from CRC32 to CRC-CCITT. Within the Utils/Hash/libcrc folder, the libcrc.c file contains the necessary functions needed to implement this new hash; all I would need to do is define the hpp/cpp files for the CRC-CCITT implementation of the Hash class(similar to the CRC32.hpp/cpp files). 

However, what is the best way to do this if I have fprime as a submodule in my project? I could copy the Utils/Hash folder into the top level of my repo, implement the files there, and include it as a subdirectory in my deployment, but then I get errors during the generate step since there are two targets with identical names. 

Anyone have any suggestions?

**Comment by Joshua-Anderson** [comment]
I think the ideal design would be to make the hash type a `#define` config option in the `config` directory, then update HashConfig to use the hash config option to select the correct hash implementation. This would require some code changes to F' itself, but is probably worth investigating for the future.

Right now though, you're right that there's no great way to do this. Ideally you be able to override the `Util/Hash` component with your implementation, but I don't believe that's supported. Two components with the same name conflict rather than override as you saw.

The easiest solution at the moment would probably be to fork the version of fprime you're using, update the hash module, and point your fprime submodule to your custom F` git fork. This isn't too difficult to do, but does make F' upgrades in the future a bit more painful. If you keep the number of custom F' patches low, even upgrading shouldn't be too bad though.

**Comment by SterlingPeet** [comment]
Is there a convenient was for cmake to patch an otherwise vanilla repository?  Thus making cmake the caretaker of vendor code patches?

Sterling

Sent from my mobile.

On Aug 24, 2022, at 8:53 PM, Joshua Anderson ***@***.***> wrote:

﻿

I think the ideal design would be to make the hash type a #define config option in the config directory, then update HashConfig to use the hash config option to select the correct hash implementation.

Right now though, you're right that there's no great way to do this. Ideally you be able to override the Util/Hash component with your implementation, but I don't believe that's supported. Two components with the same name conflict rather than override as you saw.

The easiest solution at the moment would probably be to fork the version of fprime you're using, update the hash module, and point your fprime submodule to your custom F` git fork. This isn't too difficult to do, but does make F' upgrades in the future a bit more painful. If you keep the number of custom F' patches low, even upgrading shouldn't be too bad though.

**Reply by Joshua-Anderson** [reply]
It's likely possible to do some sort of patch management in cmake, but it's not something I've seen before and I suspect that managing all patches within git itself will likely be less fragile and simpler to manage long term.

**Comment by timcanham** [comment]
The other option is to bypass the class and use the library directly if this is for a non-framework usage.
